subsys/bluetooth/mesh/access.c

```diff
@@ -506,15 +506,37 @@ static bool model_has_dst(struct bt_mesh_model *mod, uint16_t dst)
 	return mod->elem_idx == 0;
 }
 
-static const struct bt_mesh_model_op *find_op(struct bt_mesh_model *models,
-					      uint8_t model_count, uint32_t opcode,
-					      struct bt_mesh_model **model)
+static const struct bt_mesh_model_op *find_op(struct bt_mesh_elem *elem,
+					      uint32_t opcode, struct bt_mesh_model **model)
 {
 	uint8_t i;
+	uint8_t count;
+	/* This value shall not be used in shipping end products. */
+	uint16_t cid = 0xffff;
+	struct bt_mesh_model *models;
+
+	/* SIG models cannot contain 3-byte (vendor) OpCodes, and
+	 * vendor models cannot contain SIG (1- or 2-byte) OpCodes, so
+	 * we only need to do the lookup in one of the model lists.
+	 */
+	if (BT_MESH_MODEL_OP_LEN(opcode) < 3) {
+		models = elem->models;
+		count = elem->model_count;
+	} else {
+		models = elem->vnd_models;
+		count = elem->vnd_model_count;
+
+		cid = (uint16_t)(opcode & 0xffff);
+	}
+
+	for (i = 0U; i < count; i++) {
 
-	for (i = 0U; i < model_count; i++) {
 		const struct bt_mesh_model_op *op;
 
+		if (cid != 0xffff && cid != models[i].vnd.company) {
+			continue;
+		}
```

Comment:
I'm a bit worried that someone will have a use case where they're combining models from one vendor with opcodes from another, but in that case, we can add a config for disabling this optimization.